// @wecom/aibot-node-sdk/dist/index.esm.js
import { EventEmitter } from 'eventemitter3';
import axios from 'axios';
import WebSocket from 'ws';
import crypto from 'crypto';

/**
 * 消息相关类型定义
 * 按照企业微信智能机器人接收消息协议定义
 */
/** 消息类型枚举 */
var MessageType;
(function (MessageType) {
    /** 文本消息 */
    MessageType["Text"] = "text";
    /** 图片消息 */
    MessageType["Image"] = "image";
    /** 图文混排消息 */
    MessageType["Mixed"] = "mixed";
    /** 语音消息 */
    MessageType["Voice"] = "voice";
    /** 文件消息 */
    MessageType["File"] = "file";
})(MessageType || (MessageType = {}));

/**
 * API 相关类型定义
 */
/** WebSocket 命令类型常量 */
const WsCmd = {
    // ========== 开发者 → 企业微信 ==========
    /** 认证订阅 */
    SUBSCRIBE: 'aibot_subscribe',
    /** 心跳 */
    HEARTBEAT: 'ping',
    /** 回复消息 */
    RESPONSE: 'aibot_respond_msg',
    /** 回复欢迎语 */
    RESPONSE_WELCOME: 'aibot_respond_welcome_msg',
    /** 更新模板卡片 */
    RESPONSE_UPDATE: 'aibot_respond_update_msg',
    /** 主动发送消息 */
    SEND_MSG: 'aibot_send_msg',
    // ========== 企业微信 → 开发者 ==========
    /** 消息推送回调 */
    CALLBACK: 'aibot_msg_callback',
    /** 事件推送回调 */
    EVENT_CALLBACK: 'aibot_event_callback',
};
// ========== 模板卡片结构体及子结构体 ==========
/** 卡片类型枚举 */
var TemplateCardType;
(function (TemplateCardType) {
    /** 文本通知模版卡片 */
    TemplateCardType["TextNotice"] = "text_notice";
    /** 图文展示模版卡片 */
    TemplateCardType["NewsNotice"] = "news_notice";
    /** 按钮交互模版卡片 */
    TemplateCardType["ButtonInteraction"] = "button_interaction";
    /** 投票选择模版卡片 */
    TemplateCardType["VoteInteraction"] = "vote_interaction";
    /** 多项选择模版卡片 */
    TemplateCardType["MultipleInteraction"] = "multiple_interaction";
})(TemplateCardType || (TemplateCardType = {}));

/**
 * 事件相关类型定义
 */
/** 事件类型枚举 */
var EventType;
(function (EventType) {
    /** 进入会话事件：用户当天首次进入机器人单聊会话 */
    EventType["EnterChat"] = "enter_chat";
    /** 模板卡片事件：用户点击模板卡片按钮 */
    EventType["TemplateCardEvent"] = "template_card_event";
    /** 用户反馈事件：用户对机器人回复进行反馈 */
    EventType["FeedbackEvent"] = "feedback_event";
})(EventType || (EventType = {}));

/**
 * 企业微信 API 客户端
 * 仅负责文件下载等 HTTP 辅助功能，消息收发均走 WebSocket 通道
 */
class WeComApiClient {
    constructor(logger, timeout = 10000) {
        this.logger = logger;
        this.httpClient = axios.create({
            timeout,
            headers: {
                'Content-Type': 'application/json',
            },
        });
    }
    /**
     * 下载文件（返回原始 Buffer 及文件名）
     */
    async downloadFileRaw(url) {
        this.logger.info('Downloading file...');
        try {
            const response = await this.httpClient.get(url, {
                responseType: 'arraybuffer',
            });
            // 从 Content-Disposition 头中解析文件名
            const contentDisposition = response.headers['content-disposition'];
            let filename;
            if (contentDisposition) {
                // 优先匹配 filename*=UTF-8''xxx 格式（RFC 5987）
                const utf8Match = contentDisposition.match(/filename\*=UTF-8''([^;\s]+)/i);
                if (utf8Match) {
                    filename = decodeURIComponent(utf8Match[1]);
                }
                else {
                    // 匹配 filename="xxx" 或 filename=xxx 格式
                    const match = contentDisposition.match(/filename="?([^";\s]+)"?/i);
                    if (match) {
                        filename = decodeURIComponent(match[1]);
                    }
                }
            }
            this.logger.info('File downloaded successfully');
            return { buffer: Buffer.from(response.data), filename };
        }
        catch (error) {
            this.logger.error('File download failed:', error.message);
            throw error;
        }
    }
}

/**
 * 通用工具方法
 */
/**
 * 生成随机字符串
 *
 * @param length - 随机字符串长度，默认 8
 * @returns 随机字符串
 */
function generateRandomString(length = 8) {
    return crypto.randomBytes(Math.ceil(length / 2)).toString('hex').substring(0, length);
}
/**
 * 生成唯一请求 ID
 *
 * 格式：`{prefix}_{timestamp}_{random}`
 *
 * @param prefix - 前缀，通常为 cmd 名称
 * @returns 唯一请求 ID
 */
function generateReqId(prefix) {
    const timestamp = Date.now();
    const random = generateRandomString();
    return `${prefix}_${timestamp}_${random}`;
}

/** SDK 内置默认 WebSocket 连接地址 */
const DEFAULT_WS_URL = 'wss://openws.work.weixin.qq.com';
class WsConnectionManager {
    constructor(logger, heartbeatInterval = 30000, reconnectBaseDelay = 1000, maxReconnectAttempts = 10, wsUrl) {
        this.ws = null;
        this.heartbeatTimer = null;
        this.reconnectAttempts = 0;
        this.isManualClose = false;
        /** 认证凭证 */
        this.botId = '';
        this.botSecret = '';
        /** Number of consecutive missed heartbeat acks (pong) */
        this.missedPongCount = 0;
        /** Max missed pong before treating connection as dead */
        this.maxMissedPong = 2;
        /** Base delay (ms) for exponential back-off reconnection */
        this.reconnectBaseDelay = 1000;
        /** Upper cap (ms) for reconnect delay */
        this.reconnectMaxDelay = 30000;
        /** 按 req_id 分组的回复发送队列，保证同一 req_id 的消息串行发送 */
        this.replyQueues = new Map();
        /** 正在等待回执的 req_id 集合，value 包含超时定时器和 resolve/reject */
        this.pendingAcks = new Map();
        /** 回执超时时间（毫秒） */
        this.replyAckTimeout = 5000;
        /** 单个 req_id 的回复队列最大长度，超过后新消息将被拒绝 */
        this.maxReplyQueueSize = 100;
        /** 连接建立回调（WebSocket open 事件，认证尚未完成） */
        this.onConnected = null;
        /** 认证成功回调 */
        this.onAuthenticated = null;
        /** 连接断开回调 */
        this.onDisconnected = null;
        /** 收到消息回调 */
        this.onMessage = null;
        /** 重连回调 */
        this.onReconnecting = null;
        /** 错误回调 */
        this.onError = null;
        this.logger = logger;
        this.heartbeatInterval = heartbeatInterval;
        this.reconnectBaseDelay = reconnectBaseDelay;
        this.maxReconnectAttempts = maxReconnectAttempts;
        this.wsUrl = wsUrl || DEFAULT_WS_URL;
    }
    /**
     * 设置认证凭证
     */
    setCredentials(botId, botSecret) {
        this.botId = botId;
        this.botSecret = botSecret;
    }
    /**
     * 建立 WebSocket 连接（使用 SDK 内置默认地址）
     */
    connect() {
        this.isManualClose = false;
        // 清理可能未完全关闭的旧连接
        if (this.ws) {
            this.ws.removeAllListeners();
            this.ws.terminate();
            this.ws = null;
        }
        this.logger.info(`Connecting to WebSocket: ${this.wsUrl}...`);
        try {
            this.ws = new WebSocket(this.wsUrl);
            this.setupEventHandlers();
        }
        catch (error) {
            this.logger.error('Failed to create WebSocket connection:', error.message);
            this.onError?.(error);
            this.scheduleReconnect();
        }
    }
    /**
     * 设置 WebSocket 事件处理器
     */
    setupEventHandlers() {
        if (!this.ws)
            return;
        this.ws.on('open', () => {
            this.logger.info('WebSocket connection established, sending auth...');
            this.reconnectAttempts = 0;
            this.missedPongCount = 0;
            // 连接建立后立即发送认证帧
            this.sendAuth();
            this.onConnected?.();
        });
        this.ws.on('message', (data) => {
            try {
                const frame = JSON.parse(data.toString());
                this.handleFrame(frame);
            }
            catch (error) {
                this.logger.error('Failed to parse WebSocket message:', error.message);
            }
        });
        this.ws.on('close', (code, reason) => {
            const reasonStr = reason.toString() || `code: ${code}`;
            this.logger.warn(`WebSocket connection closed: ${reasonStr}`);
            this.stopHeartbeat();
            this.clearPendingMessages(`WebSocket connection closed (${reasonStr})`);
            this.onDisconnected?.(reasonStr);
            if (!this.isManualClose) {
                this.scheduleReconnect();
            }
        });
        this.ws.on('error', (error) => {
            this.logger.error('WebSocket error:', error.message);
            this.onError?.(error);
        });
        this.ws.on('ping', () => {
            this.logger.debug('Received WebSocket protocol ping, sent pong');
            this.ws?.pong();
        });
    }
    /**
     * 发送认证帧
     *
     * 格式：{ cmd: "aibot_subscribe", headers: { req_id }, body: { secret, bot_id } }
     */
    sendAuth() {
        try {
            this.send({
                cmd: WsCmd.SUBSCRIBE,
                headers: { req_id: generateReqId(WsCmd.SUBSCRIBE) },
                body: {
                    bot_id: this.botId,
                    secret: this.botSecret,
                },
            });
            this.logger.info('Auth frame sent');
        }
        catch (error) {
            this.logger.error('Failed to send auth frame:', error.message);
        }
    }
    /**
     * 处理收到的帧数据
     *
     * 接收帧结构：
     * - 消息推送：{ cmd: "aibot_msg_callback", headers: { req_id }, body: { ... } }
     * - 认证/心跳响应：{ headers: { req_id }, errcode: 0, errmsg: "ok" }
     */
    handleFrame(frame) {
        // 消息推送：cmd 为 "aibot_msg_callback"
        if (frame.cmd === WsCmd.CALLBACK) {
            this.logger.debug('Received push message:', JSON.stringify(frame.body));
            this.onMessage?.(frame);
            return;
        }
        // 事件推送：cmd 为 "aibot_event_callback"
        if (frame.cmd === WsCmd.EVENT_CALLBACK) {
            this.logger.debug('Received event callback:', JSON.stringify(frame.body));
            this.onMessage?.(frame);
            return;
        }
        // 无 cmd 的帧：认证响应、心跳响应或回复消息回执，通过 req_id 前缀区分类型，再判断 errcode
        const reqId = frame.headers?.req_id || '';
        // 检查是否是回复消息的回执（req_id 存在于 pendingAcks 中）
        if (this.pendingAcks.has(reqId)) {
            this.handleReplyAck(reqId, frame);
            return;
        }
        if (reqId.startsWith(WsCmd.SUBSCRIBE)) {
            // 认证响应
            if (frame.errcode !== 0) {
                this.logger.error(`Authentication failed: errcode=${frame.errcode}, errmsg=${frame.errmsg}`);
                this.onError?.(new Error(`Authentication failed: ${frame.errmsg} (code: ${frame.errcode})`));
                return;
            }
            this.logger.info('Authentication successful');
            this.startHeartbeat();
            this.onAuthenticated?.();
            return;
        }
        if (reqId.startsWith(WsCmd.HEARTBEAT)) {
            // 心跳响应
            if (frame.errcode !== 0) {
                this.logger.warn(`Heartbeat ack error: errcode=${frame.errcode}, errmsg=${frame.errmsg}`);
                return;
            }
            this.missedPongCount = 0;
            this.logger.debug('Received heartbeat ack');
            return;
        }
        // 未知帧类型
        this.logger.warn('Received unknown frame:', JSON.stringify(frame));
        this.onMessage?.(frame);
    }
    /**
     * 启动心跳定时器
     */
    startHeartbeat() {
        this.stopHeartbeat();
        this.heartbeatTimer = setInterval(() => {
            this.sendHeartbeat();
        }, this.heartbeatInterval);
        this.logger.debug(`Heartbeat timer started, interval: ${this.heartbeatInterval}ms`);
    }
    /**
     * 停止心跳定时器
     */
    stopHeartbeat() {
        if (this.heartbeatTimer) {
            clearInterval(this.heartbeatTimer);
            this.heartbeatTimer = null;
            this.logger.debug('Heartbeat timer stopped');
        }
    }
    /**
     * 发送心跳
     * If consecutive missed pong count reaches the threshold, treat the
     * connection as dead and trigger reconnection.
     *
     * 格式：{ cmd: "ping", headers: { req_id } }
     */
    sendHeartbeat() {
        // Check missed pong BEFORE sending the next heartbeat
        if (this.missedPongCount >= this.maxMissedPong) {
            this.logger.warn(`No heartbeat ack received for ${this.missedPongCount} consecutive pings, connection considered dead`);
            this.stopHeartbeat();
            // Force-close the underlying socket so the 'close' handler fires
            if (this.ws) {
                this.ws.terminate();
            }
            return;
        }
        this.missedPongCount++;
        try {
            this.send({
                cmd: WsCmd.HEARTBEAT,
                headers: { req_id: generateReqId(WsCmd.HEARTBEAT) },
            });
            this.logger.debug(`Heartbeat sent${this.missedPongCount > 1 ? ` (awaiting ${this.missedPongCount} pong)` : ''}`);
        }
        catch (error) {
            this.logger.error('Failed to send heartbeat:', error.message);
        }
    }
    /**
     * 安排重连
     */
    scheduleReconnect() {
        if (this.maxReconnectAttempts !== -1 && this.reconnectAttempts >= this.maxReconnectAttempts) {
            this.logger.error(`Max reconnect attempts reached (${this.maxReconnectAttempts}), giving up`);
            this.onError?.(new Error('Max reconnect attempts exceeded'));
            return;
        }
        this.reconnectAttempts++;
        // Exponential back-off: 1s, 2s, 4s, 8s … capped at reconnectMaxDelay
        const delay = Math.min(this.reconnectBaseDelay * Math.pow(2, this.reconnectAttempts - 1), this.reconnectMaxDelay);
        this.logger.info(`Reconnecting in ${delay}ms (attempt ${this.reconnectAttempts})...`);
        this.onReconnecting?.(this.reconnectAttempts);
        setTimeout(() => {
            if (this.isManualClose)
                return;
            // 重连时直接使用内置默认地址，连接建立后自动重新认证
            this.connect();
        }, delay);
    }
    /**
     * 发送数据帧
     *
     * 统一格式：{ cmd, headers: { req_id }, body }
     */
    send(frame) {
        if (this.ws && this.ws.readyState === WebSocket.OPEN) {
            this.ws.send(JSON.stringify(frame));
        }
        else {
            throw new Error('WebSocket not connected, unable to send data');
        }
    }
    /**
     * 通过 WebSocket 通道发送回复消息（串行队列版本）
     *
     * 同一个 req_id 的消息会被放入队列中串行发送：
     * 发送一条后等待服务端回执，收到回执或超时后才发送下一条。
     *
     * 格式：{ cmd: "aibot_respond_msg", headers: { req_id }, body: { ... } }
     *
     * @param reqId - 透传回调中的 req_id
     * @param body - 回复消息体（如 StreamReplyBody）
     * @param cmd - 发送的命令类型，默认 WsCmd.RESPONSE
     * @returns Promise，收到回执时 resolve(回执帧)，超时或errcode非0时 reject(Error)
     */
    sendReply(reqId, body, cmd = WsCmd.RESPONSE) {
        return new Promise((resolve, reject) => {
            const frame = {
                cmd,
                headers: { req_id: reqId },
                body,
            };
            const item = { frame, resolve, reject };
            // 入队
            if (!this.replyQueues.has(reqId)) {
                this.replyQueues.set(reqId, []);
            }
            const queue = this.replyQueues.get(reqId);
            // 防止队列无限增长导致内存泄漏
            if (queue.length >= this.maxReplyQueueSize) {
                this.logger.warn(`Reply queue for reqId ${reqId} exceeds max size (${this.maxReplyQueueSize}), rejecting new message`);
                reject(new Error(`Reply queue for reqId ${reqId} exceeds max size (${this.maxReplyQueueSize})`));
                return;
            }
            queue.push(item);
            // 如果队列中只有这一条，说明当前空闲，立即开始处理
            if (queue.length === 1) {
                this.processReplyQueue(reqId);
            }
        });
    }
    /**
     * 处理指定 req_id 的回复队列
     * 取出队列头部的消息发送，并设置回执超时
     */
    processReplyQueue(reqId) {
        const queue = this.replyQueues.get(reqId);
        if (!queue || queue.length === 0) {
            // 队列为空，清理
            this.replyQueues.delete(reqId);
            return;
        }
        const item = queue[0];
        try {
            // 发送帧
            this.send(item.frame);
            this.logger.debug(`Reply message sent via WebSocket, reqId: ${reqId}, queue length: ${queue.length}`);
        }
        catch (error) {
            this.logger.error(`Failed to send reply for reqId ${reqId}:`, error.message);
            // 发送失败：reject 当前项，继续处理队列中的下一条
            queue.shift();
            item.reject(error);
            this.processReplyQueue(reqId);
            return;
        }
        // 设置回执超时定时器
        const timer = setTimeout(() => {
            this.logger.warn(`Reply ack timeout (${this.replyAckTimeout}ms) for reqId: ${reqId}`);
            this.pendingAcks.delete(reqId);
            // 超时：reject 当前项，然后继续处理队列中的下一条
            queue.shift();
            item.reject(new Error(`Reply ack timeout (${this.replyAckTimeout}ms) for reqId: ${reqId}`));
            this.processReplyQueue(reqId);
        }, this.replyAckTimeout);
        // 注册到待回执 Map
        this.pendingAcks.set(reqId, {
            resolve: item.resolve,
            reject: item.reject,
            timer,
        });
    }
    /**
     * 处理回复消息的回执
     * 收到回执后释放队列锁，继续处理下一条
     */
    handleReplyAck(reqId, frame) {
        const pending = this.pendingAcks.get(reqId);
        if (!pending)
            return;
        // 清除超时定时器
        clearTimeout(pending.timer);
        this.pendingAcks.delete(reqId);
        const queue = this.replyQueues.get(reqId);
        if (frame.errcode !== 0) {
            this.logger.warn(`Reply ack error: reqId=${reqId}, errcode=${frame.errcode}, errmsg=${frame.errmsg}`);
            // 失败：reject 当前项
            if (queue) {
                queue.shift();
            }
            pending.reject(frame);
        }
        else {
            this.logger.debug(`Reply ack received for reqId: ${reqId}`);
            // 成功：resolve 当前项，传入完整回执帧
            if (queue) {
                queue.shift();
            }
            pending.resolve(frame);
        }
        // 继续处理队列中的下一条
        this.processReplyQueue(reqId);
    }
    /**
     * 主动断开连接
     */
    /**
     * 清理所有待处理的消息和回执
     * @param reason - 清理原因，用于 reject 的错误信息
     */
    clearPendingMessages(reason) {
        for (const [_reqId, pending] of this.pendingAcks) {
            clearTimeout(pending.timer);
        }
        this.pendingAcks.clear();
        for (const [reqId, queue] of this.replyQueues) {
            for (const item of queue) {
                item.reject(new Error(`${reason}, reply for reqId: ${reqId} cancelled`));
            }
        }
        this.replyQueues.clear();
    }
    disconnect() {
        this.isManualClose = true;
        this.stopHeartbeat();
        this.clearPendingMessages('Connection manually closed');
        if (this.ws) {
            this.ws.close(1000, 'Manual disconnect');
            this.ws = null;
        }
        this.logger.info('WebSocket connection manually closed');
    }
    /**
     * 获取当前连接状态
     */
    get isConnected() {
        return this.ws !== null && this.ws.readyState === WebSocket.OPEN;
    }
}

/**
 * 消息处理器
 * 负责解析 WebSocket 帧并分发为具体的消息事件和事件回调
 */
class MessageHandler {
    constructor(logger) {
        this.logger = logger;
    }
    /**
     * 处理收到的 WebSocket 帧，解析并触发对应的消息/事件
     *
     * WebSocket 推送帧结构：
     * - 消息推送：{ cmd: "aibot_msg_callback", headers: { req_id: "xxx" }, body: { msgid, msgtype, ... } }
     * - 事件推送：{ cmd: "aibot_event_callback", headers: { req_id: "xxx" }, body: { msgid, msgtype: "event", event: { ... } } }
     *
     * @param frame - WebSocket 接收帧
     * @param emitter - WSClient 实例，用于触发事件
     */
    handleFrame(frame, emitter) {
        try {
            const body = frame.body;
            if (!body || !body.msgtype) {
                this.logger.warn('Received invalid message format:', JSON.stringify(frame).substring(0, 200));
                return;
            }
            // 事件推送回调处理
            if (frame.cmd === WsCmd.EVENT_CALLBACK) {
                this.handleEventCallback(frame, emitter);
                return;
            }
            // 消息推送回调处理
            this.handleMessageCallback(frame, emitter);
        }
        catch (error) {
            this.logger.error('Failed to handle message:', error.message);
        }
    }
    /**
     * 处理消息推送回调 (aibot_msg_callback)
     */
    handleMessageCallback(frame, emitter) {
        const body = frame.body;
        // 触发通用消息事件
        emitter.emit('message', frame);
        // 根据 body 中的消息类型触发特定事件
        switch (body.msgtype) {
            case MessageType.Text:
                emitter.emit('message.text', frame);
                break;
            case MessageType.Image:
                emitter.emit('message.image', frame);
                break;
            case MessageType.Mixed:
                emitter.emit('message.mixed', frame);
                break;
            case MessageType.Voice:
                emitter.emit('message.voice', frame);
                break;
            case MessageType.File:
                emitter.emit('message.file', frame);
                break;
            default:
                this.logger.debug(`Received unhandled message type: ${body.msgtype}`);
                break;
        }
    }
    /**
     * 处理事件推送回调 (aibot_event_callback)
     */
    handleEventCallback(frame, emitter) {
        const body = frame.body;
        // 触发通用事件
        emitter.emit('event', frame);
        // 根据事件类型触发特定事件
        const eventType = body.event?.eventtype;
        if (eventType) {
            const eventKey = `event.${eventType}`;
            emitter.emit(eventKey, frame);
        }
        else {
            this.logger.debug('Received event callback without eventtype:', JSON.stringify(body).substring(0, 200));
        }
    }
}

/**
 * 加解密工具模块
 * 提供文件加解密相关的功能函数
 */
/**
 * 使用 AES-256-CBC 解密文件
 *
 * @param encryptedBuffer - 加密的文件数据
 * @param aesKey - Base64 编码的 AES-256 密钥
 * @returns 解密后的文件 Buffer
 */
function decryptFile(encryptedBuffer, aesKey) {
    // 参数验证
    if (!encryptedBuffer || encryptedBuffer.length === 0) {
        throw new Error('decryptFile: encryptedBuffer is empty or not provided');
    }
    if (!aesKey || typeof aesKey !== 'string') {
        throw new Error('decryptFile: aesKey must be a non-empty string');
    }
    // 将 Base64 编码的 aesKey 解码为 Buffer
    const key = Buffer.from(aesKey, 'base64');
    // IV 取 aesKey 解码后的前 16 字节
    const iv = key.subarray(0, 16);
    try {
        const decipher = crypto.createDecipheriv('aes-256-cbc', key, iv);
        // 关闭自动 padding，因为文档要求 PKCS#7 填充至 32 字节的倍数，
        // 而 Node.js 默认按 16 字节 block 去除 padding，会导致 bad decrypt 错误
        decipher.setAutoPadding(false);
        const decrypted = Buffer.concat([
            decipher.update(encryptedBuffer),
            decipher.final(),
        ]);
        // 手动去除 PKCS#7 填充（支持 32 字节 block）
        const padLen = decrypted[decrypted.length - 1];
        if (padLen < 1 || padLen > 32 || padLen > decrypted.length) {
            throw new Error(`Invalid PKCS#7 padding value: ${padLen}`);
        }
        // 验证所有 padding 字节是否一致
        for (let i = decrypted.length - padLen; i < decrypted.length; i++) {
            if (decrypted[i] !== padLen) {
                throw new Error('Invalid PKCS#7 padding: padding bytes mismatch');
            }
        }
        return decrypted.subarray(0, decrypted.length - padLen);
    }
    catch (error) {
        throw new Error(`decryptFile: Decryption failed - ${error.message}. This may indicate corrupted data or an incorrect aesKey.`);
    }
}

/**
 * 默认日志实现
 * 带有日志级别和时间戳的控制台日志
 */
class DefaultLogger {
    constructor(prefix = 'AiBotSDK') {
        this.prefix = prefix;
    }
    formatTime() {
        return new Date().toISOString();
    }
    debug(message, ...args) {
        console.debug(`[${this.formatTime()}] [${this.prefix}] [DEBUG] ${message}`, ...args);
    }
    info(message, ...args) {
        console.info(`[${this.formatTime()}] [${this.prefix}] [INFO] ${message}`, ...args);
    }
    warn(message, ...args) {
        console.warn(`[${this.formatTime()}] [${this.prefix}] [WARN] ${message}`, ...args);
    }
    error(message, ...args) {
        console.error(`[${this.formatTime()}] [${this.prefix}] [ERROR] ${message}`, ...args);
    }
}

class WSClient extends EventEmitter {
    constructor(options) {
        super();
        this.started = false;
        // 合并默认选项
        this.options = {
            reconnectInterval: 1000,
            maxReconnectAttempts: 10,
            heartbeatInterval: 30000,
            requestTimeout: 10000,
            wsUrl: '',
            logger: new DefaultLogger(),
            ...options,
        };
        this.logger = this.options.logger;
        // 初始化 API 客户端（仅用于文件下载）
        this.apiClient = new WeComApiClient(this.logger, this.options.requestTimeout);
        // 初始化 WebSocket 管理器
        this.wsManager = new WsConnectionManager(this.logger, this.options.heartbeatInterval, this.options.reconnectInterval, this.options.maxReconnectAttempts, this.options.wsUrl || undefined);
        // 设置认证凭证
        this.wsManager.setCredentials(this.options.botId, this.options.secret);
        // 初始化消息处理器
        this.messageHandler = new MessageHandler(this.logger);
        // 绑定 WebSocket 事件
        this.setupWsEvents();
    }
    /**
     * 设置 WebSocket 事件处理
     */
    setupWsEvents() {
        this.wsManager.onConnected = () => {
            this.emit('connected');
        };
        // 认证成功
        this.wsManager.onAuthenticated = () => {
            this.logger.info('Authenticated');
            this.emit('authenticated');
        };
        this.wsManager.onDisconnected = (reason) => {
            this.emit('disconnected', reason);
        };
        this.wsManager.onReconnecting = (attempt) => {
            this.emit('reconnecting', attempt);
        };
        this.wsManager.onError = (error) => {
            this.emit('error', error);
        };
        this.wsManager.onMessage = (frame) => {
            this.messageHandler.handleFrame(frame, this);
        };
    }
    /**
     * 建立 WebSocket 长连接
     * SDK 使用内置默认地址建立连接，连接成功后自动发送认证帧（botId + secret）。
     * 支持链式调用：wsClient.connect().on('message', handler)
     *
     * @returns 返回 this，支持链式调用
     */
    connect() {
        if (this.started) {
            this.logger.warn('Client already connected');
            return this;
        }
        this.logger.info('Establishing WebSocket connection...');
        this.started = true;
        // 直接使用内置默认地址建立连接，连接成功后自动认证
        this.wsManager.connect();
        return this;
    }
    /**
     * 断开 WebSocket 连接
     */
    disconnect() {
        if (!this.started) {
            this.logger.warn('Client not connected');
            return;
        }
        this.logger.info('Disconnecting...');
        this.started = false;
        this.wsManager.disconnect();
        this.logger.info('Disconnected');
    }
    /**
     * 通过 WebSocket 通道发送回复消息（通用方法）
     *
     * @param frame - 收到的原始 WebSocket 帧，透传 headers.req_id
     * @param body - 回复消息体
     * @param cmd
     */
    reply(frame, body, cmd) {
        const reqId = frame.headers?.req_id || '';
        return this.wsManager.sendReply(reqId, body, cmd);
    }
    /**
     * 发送流式文本回复（便捷方法）
     *
     * @param frame - 收到的原始 WebSocket 帧，透传 headers.req_id
     * @param streamId - 流式消息 ID
     * @param content - 回复内容（支持 Markdown）
     * @param finish - 是否结束流式消息，默认 false
     * @param msgItem - 图文混排项（仅在 finish=true 时有效），用于在结束时附带图片内容
     * @param feedback - 反馈信息（仅在首次回复时设置）
     */
    replyStream(frame, streamId, content, finish = false, msgItem, feedback) {
        const stream = {
            id: streamId,
            finish,
            content,
        };
        // msg_item 仅在 finish=true 时支持
        if (finish && msgItem && msgItem.length > 0) {
            stream.msg_item = msgItem;
        }
        // feedback 仅在首次回复时设置
        if (feedback) {
            stream.feedback = feedback;
        }
        return this.reply(frame, {
            msgtype: 'stream',
            stream,
        });
    }
    /**
     * 发送欢迎语回复
     *
     * 注意：此方法需要使用对应事件（如 enter_chat）的 req_id 才能调用，
     * 即 frame 参数应来自触发欢迎语的事件帧。
     * 收到事件回调后需在 5 秒内发送回复，超时将无法发送欢迎语。
     *
     * @param frame - 对应事件的 WebSocket 帧（需包含该事件的 req_id）
     * @param body - 欢迎语消息体（支持文本或模板卡片格式）
     */
    replyWelcome(frame, body) {
        return this.reply(frame, body, WsCmd.RESPONSE_WELCOME);
    }
    /**
     * 回复模板卡片消息
     *
     * 收到消息回调或进入会话事件后，可使用此方法回复模板卡片消息。
     *
     * @param frame - 收到的原始 WebSocket 帧，透传 headers.req_id
     * @param templateCard - 模板卡片内容
     * @param feedback - 反馈信息
     */
    replyTemplateCard(frame, templateCard, feedback) {
        const card = feedback ? { ...templateCard, feedback } : templateCard;
        const body = {
            msgtype: 'template_card',
            template_card: card,
        };
        return this.reply(frame, body);
    }
    /**
     * 发送流式消息 + 模板卡片组合回复
     *
     * 首次回复时必须返回 stream 的 id。
     * template_card 可首次回复，也可在后续回复中发送，但同一个消息只能回复一次。
     *
     * @param frame - 收到的原始 WebSocket 帧，透传 headers.req_id
     * @param streamId - 流式消息 ID
     * @param content - 回复内容（支持 Markdown）
     * @param finish - 是否结束流式消息，默认 false
     * @param options - 可选项
     * @param options.msgItem - 图文混排项（仅在 finish=true 时有效）
     * @param options.streamFeedback - 流式消息反馈信息（首次回复时设置）
     * @param options.templateCard - 模板卡片内容（同一消息只能回复一次）
     * @param options.cardFeedback - 模板卡片反馈信息
     */
    replyStreamWithCard(frame, streamId, content, finish = false, options) {
        const stream = {
            id: streamId,
            finish,
            content,
        };
        if (finish && options?.msgItem && options.msgItem.length > 0) {
            stream.msg_item = options.msgItem;
        }
        if (options?.streamFeedback) {
            stream.feedback = options.streamFeedback;
        }
        const body = {
            msgtype: 'stream_with_template_card',
            stream,
        };
        if (options?.templateCard) {
            body.template_card = options.cardFeedback
                ? { ...options.templateCard, feedback: options.cardFeedback }
                : options.templateCard;
        }
        return this.reply(frame, body);
    }
    /**
     * 更新模板卡片
     *
     * 注意：此方法需要使用对应事件（template_card_event）的 req_id 才能调用，
     * 即 frame 参数应来自触发更新的事件帧。
     * 收到事件回调后需在 5 秒内发送回复，超时将无法更新卡片。
     *
     * @param frame - 对应事件的 WebSocket 帧（需包含该事件的 req_id）
     * @param templateCard - 模板卡片内容（task_id 需跟回调收到的 task_id 一致）
     * @param userids - 要替换模版卡片消息的 userid 列表，若不填则替换所有用户
     */
    updateTemplateCard(frame, templateCard, userids) {
        const body = {
            response_type: 'update_template_card',
            template_card: templateCard,
        };
        if (userids && userids.length > 0) {
            body.userids = userids;
        }
        return this.reply(frame, body, WsCmd.RESPONSE_UPDATE);
    }
    /**
     * 主动发送消息
     *
     * 向指定会话（单聊或群聊）主动推送消息，无需依赖收到的回调帧。
     *
     * @param chatid - 会话 ID，单聊填用户的 userid，群聊填对应群聊的 chatid
     * @param body - 消息体（支持 markdown 或 template_card 格式）
     * @returns Promise，收到回执时 resolve(回执帧)
     *
     * @example
     * ```ts
     * // 发送 markdown 消息
     * await wsClient.sendMessage('CHATID', {
     *   msgtype: 'markdown',
     *   markdown: { content: '这是一条**主动推送**的消息' },
     * });
     *
     * // 发送模板卡片消息
     * await wsClient.sendMessage('CHATID', {
     *   msgtype: 'template_card',
     *   template_card: { card_type: 'text_notice', ... },
     * });
     * ```
     */
    sendMessage(chatid, body) {
        const reqId = generateReqId(WsCmd.SEND_MSG);
        const fullBody = {
            chatid,
            ...body,
        };
        return this.wsManager.sendReply(reqId, fullBody, WsCmd.SEND_MSG);
    }
    /**
     * 下载文件并使用 AES 密钥解密
     *
     * @param url - 文件下载地址
     * @param aesKey - AES 解密密钥（Base64 编码），取自消息中 image.aeskey 或 file.aeskey 字段
     * @returns 解密后的文件 Buffer 及文件名
     *
     * @example
     * ```ts
     * // aesKey 来自消息体中的 image.aeskey 或 file.aeskey
     * const { buffer, filename } = await wsClient.downloadFile(imageUrl, body.image?.aeskey);
     * ```
     */
    async downloadFile(url, aesKey) {
        this.logger.info('Downloading and decrypting file...');
        try {
            // 下载加密的文件数据
            const { buffer: encryptedBuffer, filename } = await this.apiClient.downloadFileRaw(url);
            // 如果没有提供 aesKey，直接返回原始数据
            if (!aesKey) {
                this.logger.warn('No aesKey provided, returning raw file data');
                return { buffer: encryptedBuffer, filename };
            }
            // 使用独立的解密模块进行 AES-256-CBC 解密
            const decryptedBuffer = decryptFile(encryptedBuffer, aesKey);
            this.logger.info('File downloaded and decrypted successfully');
            return { buffer: decryptedBuffer, filename };
        }
        catch (error) {
            this.logger.error('File download/decrypt failed:', error.message);
            throw error;
        }
    }
    /**
     * 获取当前连接状态
     */
    get isConnected() {
        return this.wsManager.isConnected;
    }
    /**
     * 获取 API 客户端实例（供高级用途使用，如文件下载）
     */
    get api() {
        return this.apiClient;
    }
}

/** 默认导出 AiBot 命名空间 */
const AiBot = {
    WSClient,
};

export { DefaultLogger, EventType, MessageHandler, MessageType, TemplateCardType, WSClient, WeComApiClient, WsCmd, WsConnectionManager, decryptFile, AiBot as default, generateRandomString, generateReqId };
